stdlib/runtime/runtimeDebug.gr

```diff
@@ -0,0 +1,712 @@
+import WasmI32 from "runtime/unsafe/wasmi32"
+import Conv from "runtime/unsafe/conv"
+import Memory from "runtime/unsafe/memory"
+import GC from "runtime/gc"
+import { coerceNumberToWasmI32 } from "runtime/numbers"
+import { allocateString } from "runtime/dataStructures"
+import NumberUtils from "runtime/numberUtils"
+import Map from "map"
+import Set from "set"
+import Option from "option"
+import Tags from "runtime/unsafe/tags"
+
+@disableGC
+let rec toHex = (n: Number) => {
+  let x = coerceNumberToWasmI32(n)
+  let result = NumberUtils.itoa32(x, 16n)
+
+  Memory.decRef(WasmI32.fromGrain(n))
+  Memory.decRef(WasmI32.fromGrain(toHex))
+
+  result
+}
+
+export enum LeakDebugSummaryMode {
+  AlwaysPrintSummary,
+  PrintSummaryIfLeaksAreFound,
+}
+
+export record LeakDebugBlockOptions {
+  printIncrefsAndDecRefs: Bool,
+  printMallocAndFree: Bool,
+  printObjectContents: Bool,
+  summary: LeakDebugSummaryMode,
+}
+
+let incrementCountInMap = (key: k, map: Map.Map<k, Number>) => {
+  match (Map.get(key, map)) {
+    Some(cnt) => Map.set(key, cnt + 1, map),
+    None => Map.set(key, 1, map),
+  }
+}
+
+let keyUnionOfTwoMaps = (map1: Map.Map<k, v1>, map2: Map.Map<k, v2>) => {
+  let keyUnion = Set.make(): (Set.Set<k>)
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map1,
+  )
+
+  Map.forEach(
+    (k, v) => {
+      Set.add(k, keyUnion)
+    },
+    map2,
+  )
+
+  keyUnion
+}
+
+let leakDebugBlockImpl = (options: LeakDebugBlockOptions, f: () -> Void) => {
+  // The main objective is to find and print objects allocated
+  // in the function, but not deallocated at the end.
+
+  let allocations = Set.make(): (Set.Set<Number>)
+
+  let tagToTypeName = (tag: Number) => {
+    match (tag) {
+      1 => "string",
```

Comment:
Maybe we should use the constants from `runtime/unsafe/tags`?

Reply:
I didn't use them to avoid `@disableGC`. Should I convert `tagToTypeName` to no gc?

Reply:
I hadn't considered that! In that case, I think a comment is fine for now (maybe also add a comment to `tags.gr` so we can remember to keep this in sync?).

Reply:
I changed this function to use `@disableGC`. It's not pretty, but probably still better than not using the constants.